service/raft/raft_group_registry.cc

```diff
@@ -406,6 +406,8 @@ raft_group_registry::~raft_group_registry() = default;
 future<bool> direct_fd_pinger::ping(direct_failure_detector::pinger::endpoint_id id, abort_source& as) {
     auto addr = _address_map.find(raft::server_id{id});
     if (!addr) {
+        static thread_local logger::rate_limit rate_limit(std::chrono::minutes(5));
```

Comment:
In the issue I specified:
> Also, warnings corresponding to different endpoints should be rate-limited separately.

However this will apply a single rate_limit to all endpoints.

If we have two endpoints with missing addresses, and get a warning for one, then for the other within 5 minutes, we will not learn that the second endpoint is also missing an address.

We should have a separate `rate_limit` for each endpoint.

Perhaps we could achieve that using something like `static thread_local std::unordered_map<direct_failure_detector::pinger::endpoint_id, logger:rate_limit>`?

Reply:
Thank you for your quick response! You are right!

I am trying to understand do we need a tread local storage to store an `unordered_map<endpoint_id, rate_limit>` and if it is thread-safe to have it as member in the `direct_fd_pinger` class:
```
class direct_fd_pinger : /* ... */ public direct_failure_detector::pinger
{
// ...
using rate_limits = std::unordered_map<direct_failure_detector::pinger::endpoint_id, logger::rate_limit>;
rate_limits _rate_limits;
// ..

future<bool> ping(direct_failure_detector::pinger::endpoint_id id, abort_source& as) override;
};
```
and than in `direct_fd_pinger::ping(..)`
```C++
auto [it, _] = _rate_limits.try_emplace(id, std::chrono::minutes(5));
auto& rate_limit = it->second;
```

`direct_failure_detector::failure_detector` 1 <-> N `direct_failure_detector::endpoint_worker`
`endpoint_worker::ping_fiber(..)` -> `ping_with_timeout(..)` -> `pinger::ping(..)` (class `direct_fd_pinger` implements the `pinger` interface)

As there is a 1 <-> N relationship between the `failure_detector` and the `endpoint_worker`. Each the `endpoint_worker` has access to the `direct_fd_pinger` through the parent [failure_detector.cc#L520](https://github.com/scylladb/scylladb/blob/681bd62424cd4a3ff1031d2a8f477b7742f2aebd/direct_failure_detector/failure_detector.cc#L520)

Could I assume that all `endpoint_worker` are executed in the same thread and not need to guard `unordered_map<endpoint_id, rate_limit>`?

Reply:
`endpoint_worker`s are running on different shards, but that's fine.

There is a separate instance of `direct_fd_pinger` on each shard:
```
            static sharded<service::direct_fd_pinger> fd_pinger;
            supervisor::notify("starting direct failure detector pinger service");
            fd_pinger.start(std::ref(messaging), std::ref(raft_address_map)).get();
```
(from main.cc)

 A given `endpoint_worker` will only access the pinger running on the same shard as the worker. And recall that there is exactly one thread per shard (Seastar shared-nothing architecture).

So you can place the `_rate_limits` map inside `direct_fd_pinger` and access it in `ping`, and you don't need any synchronization mechanisms.

Reply:
Updated